### PR TITLE
Use canonical status code mapping from HTTP to RPC

### PR DESCRIPTION
**Pull request by jskeet**
This is now *not* consistent with [the Java implementation](https://github.com/googleapis/gax-java/blob/main/gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpJsonStatusCode.java) in some cases.

This is an alternative to #490

### CODE REVIEW

File: Google.Api.Gax.Grpc/Rest/RestChannel.cs [hunk 1]
```diff
@@ -186,16 +186,36 @@ internal Metadata GetHeaders()
...
-                    HttpStatusCode.OK => StatusCode.OK,
-                    HttpStatusCode.Unauthorized => StatusCode.Unauthenticated,
-                    HttpStatusCode.Forbidden => StatusCode.PermissionDenied,
-                    HttpStatusCode.BadRequest => StatusCode.InvalidArgument,
-                    HttpStatusCode.InternalServerError => StatusCode.Internal,
-                    HttpStatusCode.NotFound => StatusCode.NotFound,
-                    HttpStatusCode.Conflict => StatusCode.FailedPrecondition,
-                    // TODO: Others!
+                    // All 2xx responses map to OK
+                    2 => StatusCode.OK,
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/RestChannel.cs:193]: +1 to this.

The Apiary libraries marks all 2xx as success, if we didn't do this here, we could get the exact same operation succeed in Apiary but "fail" in REGAPIC.

File: Google.Api.Gax.Grpc/Rest/RestChannel.cs [hunk 2]
```diff
@@ -186,16 +186,36 @@ internal Metadata GetHeaders()
...
+                    // 5xx defaults to Internal, with specific exceptions
+                    5 => httpStatus switch
+                    {
+                        501 => StatusCode.Unimplemented,
+                        503 => StatusCode.Unavailable,
+                        504 => StatusCode.DeadlineExceeded,
+                        _ => StatusCode.Internal
+                    },
+
+                    // Anything else (including all 1xx and 3xx) maps to Unknown
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/RestChannel.cs:218]: We handle redirects in Apiary.

Comment (amanda-tarafa) [reply Google.Api.Gax.Grpc/Rest/RestChannel.cs:218]: I have no idea if they are used, tbh. I've never seen an actual redirect while debugging code, but of course, I've onle ever debugged a low percent of operations. Maybe we can at least chat to the Compute API folk and ask, or maybe to ESF?

Comment (vchudnov-g) [reply Google.Api.Gax.Grpc/Rest/RestChannel.cs:218]: Any progress on the redirect issue? I would assume we'd want to handle it like Apiary and just follow the redirect. Ideally, that means HttpClient would do this and redirection codes would not show up here in practice....

Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/RestChannel.cs:218]: Sorry, I haven't chased the redirect issue at all. We could definitely "ask" HttpClient to do it for us, although I don't know whether it gives up after a few redirects.

Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/RestChannel.cs:218]: I suggest we can merge this change without worrying about redirects for now, but I'll file an issue to make sure we remember.

**Comment by jskeet** [comment]
This has now been agreed internally, and is ready to merge if you're happy, @amanda-tarafa

File: Google.Api.Gax.Grpc/Rest/RestChannel.cs (hunk 2 above)
Comment (amanda-tarafa) [reply Google.Api.Gax.Grpc/Rest/RestChannel.cs:218]: Just for completeness, this is being tracked in #511.

**Review by amanda-tarafa — APPROVED** [review]
LGTM

**Pull request merged**